sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java

```diff
@@ -1061,51 +1066,28 @@ private <T> Flux<FeedResponse<T>> createQueryInternal(
         }, Queues.SMALL_BUFFER_SIZE, 1);
     }
 
-    private static void applyExceptionToMergedDiagnostics(
-        CosmosQueryRequestOptions requestOptions,
-        CosmosException exception) {
+    private static void applyExceptionToMergedDiagnosticsForQuery(
+        CosmosException exception,
+        DiagnosticsClientContext diagnosticsClientContext) {
 
         List<CosmosDiagnostics> cancelledRequestDiagnostics =
-            qryOptAccessor
-                .getCancelledRequestDiagnosticsTracker(requestOptions);
+            diagnosticsClientContext.getCreatedDiagnostics();
 
-        // if there is any cancelled requests, collect cosmos diagnostics
         if (cancelledRequestDiagnostics != null && !cancelledRequestDiagnostics.isEmpty()) {
-            // combine all the cosmos diagnostics
-            CosmosDiagnostics aggregratedCosmosDiagnostics =
-                cancelledRequestDiagnostics
-                    .stream()
-                    .reduce((first, toBeMerged) -> {
-                        ClientSideRequestStatistics clientSideRequestStatistics =
-                            ImplementationBridgeHelpers
-                                .CosmosDiagnosticsHelper
-                                .getCosmosDiagnosticsAccessor()
-                                .getClientSideRequestStatisticsRaw(first);
-
-                        ClientSideRequestStatistics toBeMergedClientSideRequestStatistics =
-                            ImplementationBridgeHelpers
-                                .CosmosDiagnosticsHelper
-                                .getCosmosDiagnosticsAccessor()
-                                .getClientSideRequestStatisticsRaw(first);
-
-                        if (clientSideRequestStatistics == null) {
-                            return toBeMerged;
-                        } else {
-                            clientSideRequestStatistics.mergeClientSideRequestStatistics(toBeMergedClientSideRequestStatistics);
-                            return first;
-                        }
-                    })
-                    .get();
-
-            BridgeInternal.setCosmosDiagnostics(exception, aggregratedCosmosDiagnostics);
+            // When reaching here, it means the query(s) has timed out based on the e2e timeout config policy
+            // Since all the underlying ongoing query requests will all timed out
+            // We just use the last cosmosDiagnostics in the scoped diagnostics factory to populate the exception
+            BridgeInternal.setCosmosDiagnostics(
```

Comment:
previously we are using the cancelledRequestDiagnostics tracked on the query options to populate the diagnostics for the exception, but I do not think it is a good approach, so changed to use scopedDiagnosticsFactory. But when using scopedDiagnosticsFactory, since it can contain multiple diagnostics, so the question will be pick which one to populate the exception. 

why the current approach has issue:
1. When the operation timed out,  there may not be any cancelled pending requests due to backoff. In this case no diagnostics will be populated in the diagnostics, we will lose all diagnostics we have 
2. For query, e2eTimeoutPolicy sits on top of availabilityStratey, so the cancelled cosmos diagnostics may also coming from hedging regions. Simply merging them will not be right

Reply:
I believe we need to revert this refactoring - it results in several tests ending up with CosmosDiagnostics being null

Reply:
Fixed in my version of this PR (did not have rights to change your branch)